crypto/provider_core.c

```diff
@@ -521,10 +524,10 @@ static int provider_init(OSSL_PROVIDER *prov)
             }
 
             if ((store = get_provider_store(prov->libctx)) == NULL
-                    || !CRYPTO_THREAD_read_lock(store->lock))
+                    || !CRYPTO_THREAD_read_lock(store->default_path_lock))
                 goto end;
-            load_dir = store->default_path;
-            CRYPTO_THREAD_unlock(store->lock);
+            load_dir = OPENSSL_strdup(store->default_path);
```

Comment:
This is wrong :(

First up, what if `store->defaultpath == NULL`?  This will often `strdup(3)` and return a non-NULL pointer.

Secondly, there needs to be a check for the `OPENSSL_strdup` failing.

I'd expect something like:
``` C
if (store->default_path != NULL) {
    load_dir = OPENSSL_strdup(store->default_path);
    if (load_dir == NULL) {
        CRYPTO_THREAD_unock(store->default_path_lock);
        return 0;
    }
}
```